### PR TITLE
Add ability to filter rewards between epochs

### LINKED ISSUE #6650: Historic reward summary per epoch
# Feature Overview

**In order to** provide an overview of the rewards earned 
**We will** provide a way to query reward types over a specified window of epochs
**So that** there is a way to accurately present aggregated reward information to the user

Example for an API that would facilitate this, not set in stone if this doesn't make sense data node side, but filtering by epoch would enable this really easily, as suggested by Phil.

```
query PartyRewards {
    party (id: 'foo') {
        rewards ( fromEpoch: 1, toEpoch: 10) {
           pagination // as normal
           edges {
                 asset
                 amount
                 rewardType
                 marketId
                 // ....ect/
              }
           }
        }
    }
}
```

# Background

Currently we cannot query reward types per epoch, only the last N rewards and therefore have no way to accurately present to the user the rewards they have earned per epoch.

Screenshot 2022-10-31 at 16 18 29

Implementing this solution with existing API would mean we take a risk on presenting an epoch with an incomplete data set - the totals could be wrong, and it's possible we miss out whole reward types as well as individual rewards - and we'd have no way of knowing it.

# Tasks
- [ ]
- [ ]

# Acceptance Criteria
A list of criteria (aim for 3!) that have to be met for this feature to be accepted as 'done' by the product owner.

- It is possible to

Acceptance Criteria and Feature Test Scenarios can, in some cases, be closely related. If acceptance criteria become fully covered by feature test scenarios they can be removed leaving on acceptance criteria that can't be directly or fully proven with tests.

# Test Scenarios
Detailed scenarios (1-3!) that can be executed as feature tests to verify that the feature has been implemented as expected. We use the follow format:

GIVEN (setup/context) 
WHEN (action) 
THEN (assertion) For example...
See [here](https://github.com/vegaprotocol/vega/tree/develop/integration/) for more format information and examples.

# Additional Details (optional)
Any additional information that provides context or gives information that will help us develop the feature. 

# Examples (optional)
Code snippets from the spec for reference

# Definition of Done
>ℹ️ Not every issue will need every item checked, however, every item on this list should be properly considered and actioned to meet the [DoD](https://github.com/vegaprotocol/vega/blob/develop/DEFINITION_OF_DONE.md).

**Before Merging**
- [ ] Create relevant for [system-test](https://github.com/vegaprotocol/system-tests/issues) tickets with feature labels
- [ ] Code refactored to meet SOLID and other code design principles
- [ ] Code is compilation error, warning, and hint free
- [ ] Carry out a basic happy path end-to-end check of the new code
- [ ] All APIs are documented so auto-generated documentation is created
- [ ] All acceptance criteria confirmed to be met, or, reasons why not discussed with the engineering leadership team
- [ ] All Unit, Integration and BVT tests are passing
- [ ] Implementation is peer reviewed (coding standards, meeting acceptance criteria, code/design quality)
- [ ] Create [front end](https://github.com/vegaprotocol/token-frontend/issues) or [console](https://github.com/vegaprotocol/console/issues) tickets with feature labels (should be done when starting the work if dependencies known i.e. API changes)

**After Merging**
- [ ] Move development ticket to `Done` if there is **NO** requirement for new system-tests
- [ ] Resolve any issues with broken system-tests
- [ ] Create [documentation](https://github.com/vegaprotocol/documentation/issues) tickets with feature labels if functionality has changed, or is a new feature

### PR DESCRIPTION
**Pull request by pscott31**
Closes #6650

### CODE REVIEW

**Review by dexturr — DISMISSED** [review]

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -2029,6 +2029,10 @@ type Party {
     assetId: ID
     "Optional pagination information"
     pagination: Pagination
+    "Optionally only return rewards >= this epoch"
+    fromEpoch: Int
+    "Optionally only return rewards <= this epoch"
+    toEpoch: Int
```
Comment (edd) [inline datanode/gateway/graphql/schema.graphql:2035]: **Nitpick (non-blocking)**: Ultra nitpick, but this is a uint64 in the GRPC, so I guess it’s technically possible that this *will eventually* overflow, if the epoch is set unreasonably short and the network lives for a long time. Hopefully by that point all browsers properly support big ints..

Regardless, I’m happy seeing an Int and not another string 😄

**Review by edd — DISMISSED** [review]

File: datanode/gateway/graphql/schema.graphql (hunk 1 above)
Comment (pscott31) [reply datanode/gateway/graphql/schema.graphql:2035]: yes fair enough, we use Int elsewhere for epoch though so at least it's consistent.. I'm going to just pretend it's never going to happen for now :)

**Review by edd — APPROVED** [review]

**Review by dexturr — APPROVED** [review]

File: datanode/gateway/graphql/resolvers.go [hunk 2]
```diff
@@ -1181,11 +1181,30 @@ func (r *myPartyResolver) TransfersConnection(
 	return r.r.transfersConnection(ctx, &party.Id, direction, pagination)
 }
 
-func (r *myPartyResolver) RewardsConnection(ctx context.Context, party *types.Party, assetID *string, pagination *v2.Pagination) (*v2.RewardsConnection, error) {
+func (r *myPartyResolver) RewardsConnection(ctx context.Context, party *types.Party, assetID *string, pagination *v2.Pagination, fromEpoch *int, toEpoch *int) (*v2.RewardsConnection, error) {
```
Comment (zale144) [inline datanode/gateway/graphql/resolvers.go:1184]: Maybe if pagination would be the last parameter would be a bit cleaner, although I'd bundled these (party, assetID, fromEpoch, toEpoch) into a single filter struct. No biggie

File: datanode/sqlstore/rewards_test.go [hunk 3]
```diff
@@ -702,7 +671,7 @@ func testRewardsCursorPaginationLastPageBeforeNewestFirst(t *testing.T) {
 	before := entities.NewCursor(entities.RewardCursor{PartyID: partyID, AssetID: assetID, EpochID: 643}.String()).Encode()
 	pagination, err := entities.NewCursorPagination(nil, nil, &last, &before, true)
 	require.NoError(t, err)
-	got, pageInfo, err := rs.GetByCursor(ctx, &partyID, &assetID, pagination)
+	got, pageInfo, err := rs.GetByCursor(ctx, &partyID, &assetID, nil, nil, pagination)
```
Comment (zale144) [inline datanode/sqlstore/rewards_test.go:674]: Yeah, with a filter struct we wouldn't have these `nil`s everywhere :)

**Review by zale144 — APPROVED** [review]
Some minor comments on aesthetics, but lgtm

**Pull request merged**